presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -15,101 +15,144 @@
-        if (!queue.isEmpty() || finished.get()) {
-            notEmptyFuture.get().complete(null);
+        if (futureToSignal != null) {
+            futureToSignal.complete(null);
+        }
+    }
+
+    public CompletableFuture<?> offer(T element)
+    {
+        requireNonNull(element);
+
+        CompletableFuture<?> result;
+        CompletableFuture<?> futureToSignal = null;
+        synchronized (this) {
+            if (finishing) {
+                return COMPLETED_FUTURE_OF_NULL;
+            }
+            elements.add(element);
+            int newSize = elements.size();
+            if (newSize == 1) {
+                futureToSignal = notEmptySignal;
+                notEmptySignal = new CompletableFuture<>();
+            }
+            result = newSize < softMaxQueueSize ? COMPLETED_FUTURE_OF_NULL : notFullSignal;
         }
 
-        return ImmutableList.copyOf((Collection<T>) elements);
+        if (futureToSignal != null) {
+            futureToSignal.complete(null);
+        }
+        return result;
     }
 
-    /**
-     * Finishes the queue. Elements added after the queue is finished
-     * are ignored without error.
-     */
-    public void finish()
+    private List<T> getBatch(int maxSize)
     {
-        finished.set(true);
-        if (queue.isEmpty()) {
-            notEmptyFuture.get().complete(null);
+        List<T> result;
+        CompletableFuture<?> futureToSignal = null;
+        synchronized (this) {
+            int oldSize = elements.size();
+            int reduceBy = Math.min(maxSize, oldSize);
+            if (reduceBy == 0) {
+                return ImmutableList.of();
+            }
+            result = new ArrayList<>(reduceBy);
+            for (int i = 0; i < reduceBy; i++) {
+                result.add(elements.remove());
+            }
+            // This checks that the queue size changed from above threshold to below. Therefore, writers shall be notified.
+            if (oldSize >= softMaxQueueSize && oldSize - reduceBy < softMaxQueueSize) {
+                futureToSignal = notFullSignal;
+                notFullSignal = new CompletableFuture<>();
+            }
+        }
+
+        if (futureToSignal != null) {
+            futureToSignal.complete(null);
         }
+        return result;
     }
 
-    /**
-     * Is the queue finished and no more elements can be fetched?
-     */
-    public boolean isFinished()
+    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
     {
-        return finished.get() && queue.isEmpty();
+        checkArgument(maxSize >= 0, "maxSize must be at least 0");
+        synchronized (this) {
+            List<T> list = getBatch(maxSize);
+            if (list.isEmpty()) {
+                if (finishing) {
+                    return completedFutureOfEmptyList;
+                }
+                return notEmptySignal.thenApply(x -> getBatch(maxSize));
```

Comment:
this must be added outside of the lock because if the future is already complete it will execute immediately